File: crates/lanekeep-rules/rules/no-unused-exports.ts
import { defineRule } from 'lanekeep'
import { resolveImport } from 'lanekeep/paths'

/**
 * Exports nobody in the corpus imports.
 *
 * Dead exported code is worse than dead private code: it is part of a module's surface, so
 * a reader has to assume something depends on it, and an agent will happily wire new code
 * up to it. Nothing in the exporting file reveals the problem — which is exactly why this
 * needs the reduce phase rather than a query.
 *
 * Matching is by `(resolved module, symbol)`, not by symbol name. A rule that treated any
 * `parse` imported anywhere as covering every exported `parse` would go quiet on a codebase
 * with common names, which is every codebase.
 *
 * @example
 * ```ts
 * import noUnusedExports from 'lanekeep/no-unused-exports'
 *
 * export default defineConfig({
 *   rules: [noUnusedExports({ entryPoints: ['src/index.ts', 'src/cli.ts'] })],
 * })
 * ```
 */
export default function noUnusedExports(options) {
  const entryPoints = options?.entryPoints ?? []

  return defineRule({
    id: 'lanekeep/no-unused-exports',
    severity: 'error',

    card: {
      message: 'unused export',
      remediation:
        'stop exporting it, or delete it — an export nobody imports is surface with no users',
      examples: {
        bad: 'export function helper() {}',
        good: 'function helper() {}',
      },
    },

    // The whole corpus feeds one pass, so the default per-invocation budget is the wrong
    // shape for it. See docs/cross-file-rules.md.
    timeout: 5_000,

    query: `
      (export_statement
        declaration: (function_declaration name: (identifier) @name)) @stmt

      (export_statement
        declaration: (class_declaration name: (type_identifier) @name)) @stmt

      (export_statement
        declaration: (lexical_declaration
          (variable_declarator name: (identifier) @name))) @stmt

      (export_statement
        (export_clause
          (export_specifier !alias name: (identifier) @name))) @stmt

      (export_statement
        (export_clause
          (export_specifier alias: (identifier) @name))) @stmt

      ; Children match in tree order, and \`import_clause\` precedes \`source\` in the
      ; grammar — the other way round is a pattern that can never match.
      ; The *name*, aliased or not: \`import { x as y }\` consumes \`x\` from the target.
      ; Capturing the alias would leave \`x\` looking unused and report it.
      (import_statement
        (import_clause (named_imports (import_specifier name: (identifier) @imported)))
        source: (string) @source)

      (import_statement
        (import_clause (namespace_import))
        source: (string) @source) @wildcard

      (export_statement source: (string) @source) @wildcard
    `,

    check(ctx, m) {
      if (m.wildcard) {
        // `import * as ns from './a'` and `export * from './a'` both consume the whole
        // module, and neither names what it takes. Treating them as covering everything in
        // the target is the only sound reading — the alternative is reporting exports that
        // are demonstrably reachable.
        ctx.emitFact({ kind: 'wildcard', from: unquote(ctx.text(m.source)) })
        return
      }

      if (m.imported) {
        ctx.emitFact({
          kind: 'import',
          symbol: ctx.text(m.imported),
          from: unquote(ctx.text(m.source)),
        })
        return
      }

      ctx.emitFact({
        kind: 'export',
        symbol: ctx.text(m.name),
        line: ctx.line(m.stmt),
        column: ctx.column(m.stmt),
      })
    },

    reduce(ctx) {
      const files = new Set(ctx.files)

      // Anything a wildcard reaches is used, whatever its name.
      const consumedWhole = new Set()
      for (const w of ctx.facts('wildcard')) {
        const target = resolveImport(w.file, w.from, files)
        if (target) consumedWhole.add(target)
      }

      // Keyed `${file}\u0000${symbol}`. A NUL separator rather than a space, because a
      // path may contain spaces and `a b` + `c` would then key the same as `a` + `b c`.
      // Written as an escape: a raw NUL in source is invisible in an editor and makes
      // the parser report a failure lines away from the character that caused it.
      const used = new Set()
      for (const i of ctx.facts('import')) {
        const target = resolveImport(i.file, i.from, files)
        if (target) used.add(`${target}\u0000${i.symbol}`)
      }

      for (const e of ctx.facts('export')) {
        // An entry point is exported for something outside the corpus — a bin, a package
        // main, a test runner. Without this the rule reports every public API of a library
        // as unused, which is true of the corpus and useless as advice.
        if (entryPoints.includes(e.file)) continue
        if (consumedWhole.has(e.file)) continue
        if (used.has(`${e.file}\u0000${e.symbol}`)) continue

        ctx.report(
          { file: e.file, line: e.line, column: e.column },
          `'${e.symbol}' is exported but nothing imports it`,
        )
      }
    },
  })
}

/** Strip the quotes from a string literal as it appears in source. */
function unquote(raw) {
  return raw.slice(1, -1)
}

File: crates/lanekeep-cli/tests/no_unused_exports.rs
//! `lanekeep/no-unused-exports`, run through the real engine.
//!
//! A cross-file rule needs more than one file, so these build a project directly rather
//! than going through `RuleTester`, which runs one subject at a time.

#![expect(
    clippy::expect_used,
    clippy::panic,
    reason = "`clippy.toml`'s allow-*-in-tests only reaches `#[test]` functions and \
              `#[cfg(test)]` modules. The `corpus` helpers are neither, so the grant it \
              already makes for unit tests has to be restated for them."
)]

mod corpus;

use corpus::Corpus;

fn corpus(options: &str, files: &[(&str, &str)]) -> Corpus {
    Corpus::new("no-unused-exports", options, files)
}

#[test]
fn an_imported_export_is_not_reported() {
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export function used() {}\n"),
            ("src/b.ts", "import { used } from './a';\nused();\n"),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn an_export_nobody_imports_is_reported() {
    let found = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "export function used() {}\nexport function orphan() {}\n",
            ),
            ("src/b.ts", "import { used } from './a';\nused();\n"),
        ],
    )
    .run();
    assert_eq!(
        found,
        vec!["src/a.ts:2:1 'orphan' is exported but nothing imports it"]
    );
}

#[test]
fn a_symbol_of_the_same_name_from_another_module_does_not_count() {
    // The reason matching is by (module, symbol) rather than by name. `parse` is a common
    // name; a rule that matched on it alone would go quiet on any real codebase.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export function parse() {}\n"),
            ("src/b.ts", "export function parse() {}\n"),
            ("src/c.ts", "import { parse } from './b';\nparse();\n"),
        ],
    )
    .run();
    assert_eq!(
        found,
        vec!["src/a.ts:1:1 'parse' is exported but nothing imports it"]
    );
}

#[test]
fn classes_and_constants_are_covered_too() {
    let found = corpus(
        "{}",
        &[(
            "src/a.ts",
            "export class Parser {}\nexport const VERSION = 1;\n",
        )],
    )
    .run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'Parser' is exported but nothing imports it",
            "src/a.ts:2:1 'VERSION' is exported but nothing imports it",
        ]
    );
}

#[test]
fn an_import_without_an_extension_resolves() {
    // `./a` has to find `a.ts`. Without this the rule reports everything.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export const x = 1;\n"),
            ("src/b.ts", "import { x } from './a';\nconsole.log(x);\n"),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_directory_import_resolves_to_its_index() {
    let found = corpus(
        "{}",
        &[
            ("src/thing/index.ts", "export const x = 1;\n"),
            (
                "src/b.ts",
                "import { x } from './thing';\nconsole.log(x);\n",
            ),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn an_import_that_walks_up_resolves() {
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export const x = 1;\n"),
            (
                "src/deep/b.ts",
                "import { x } from '../a';\nconsole.log(x);\n",
            ),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_namespace_import_covers_everything_in_the_target() {
    // `import * as a` does not name what it takes, so anything in `a` is demonstrably
    // reachable. Reporting those would be plainly wrong.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export const x = 1;\nexport const y = 2;\n"),
            ("src/b.ts", "import * as a from './a';\nconsole.log(a);\n"),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_star_re_export_covers_everything_in_the_target() {
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export const x = 1;\nexport const y = 2;\n"),
            ("src/index.ts", "export * from './a';\n"),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn an_entry_point_is_exempt() {
    // A library's public API is imported from outside the corpus. Without this the rule
    // reports every exported symbol of every package, which is true and useless.
    let found = corpus(
        "{ entryPoints: ['src/index.ts'] }",
        &[("src/index.ts", "export function publicApi() {}\n")],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_non_entry_point_is_still_reported_when_entry_points_are_configured() {
    let found = corpus(
        "{ entryPoints: ['src/index.ts'] }",
        &[
            ("src/index.ts", "export function publicApi() {}\n"),
            ("src/internal.ts", "export function hidden() {}\n"),
        ],
    )
    .run();
    assert_eq!(
        found,
        vec!["src/internal.ts:1:1 'hidden' is exported but nothing imports it"]
    );
}

#[test]
fn a_package_import_is_ignored_rather_than_resolved() {
    // `lodash` names nothing in the corpus. The export below is still unused, and the
    // unresolvable import must not accidentally mark it used.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export const merge = 1;\n"),
            (
                "src/b.ts",
                "import { merge } from 'lodash';\nconsole.log(merge);\n",
            ),
        ],
    )
    .run();
    assert_eq!(
        found,
        vec!["src/a.ts:1:1 'merge' is exported but nothing imports it"]
    );
}

#[test]
fn an_aliased_import_is_matched_on_the_exported_name() {
    // `import { x as y }` consumes `x`. Matching on `y` would report `x` as unused.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export const x = 1;\n"),
            (
                "src/b.ts",
                "import { x as y } from './a';\nconsole.log(y);\n",
            ),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn an_aliased_export_is_reported_under_the_name_it_publishes() {
    let found = corpus(
        "{}",
        &[(
            "src/a.ts",
            "const internal = 1;\nexport { internal as published };\n",
        )],
    )
    .run();
    assert_eq!(
        found,
        vec!["src/a.ts:2:1 'published' is exported but nothing imports it"]
    );
}

#[test]
fn the_same_corpus_reports_the_same_thing_every_run() {
    let corpus = corpus(
        "{}",
        &[
            ("src/a.ts", "export const a1 = 1;\nexport const a2 = 2;\n"),
            ("src/b.ts", "export const b1 = 1;\nexport const b2 = 2;\n"),
            ("src/c.ts", "export const c1 = 1;\nexport const c2 = 2;\n"),
            ("src/d.ts", "import { a1 } from './a';\nconsole.log(a1);\n"),
        ],
    );
    let first = corpus.run();
    assert!(!first.is_empty(), "the fixture should report something");
    for attempt in 0..4 {
        assert_eq!(corpus.run(), first, "output changed on attempt {attempt}");
    }
}
